lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -242,148 +229,57 @@ define((require, exports, module) ->
-        { type: "variable.parameter", value: "equation" }
-        { type: "paren.rparen", value: "}" }
-      ]
-      [
-        { type: "storage.type", value: "\\end" }
-        { type: "paren.lparen", value: "{" }
-        { type: "variable.parameter", value: "equation*" }
-        { type: "paren.rparen", value: "}" }
-      ]
-      [ { type: "string", value: "\\]" } ]
-      [ { type: "string", value: "\\)" } ]
-      [ { type: "string", value: "$" } ]
-      [ { type: "string", value: "$$" } ]
-    ]
+    @DOCUMENT_END_ERROR_CODE: "js.math_preview.error.document_end"
+    @EMPTY_LINE_ERROR_CODE: "js.math_preview.error.empty_line"
+    @WHITESPACE_LINE_ERROR_CODE: "js.math_preview.error.whitespace_line"
 
     # empty constructor
     constructor: (@editor) ->
 
-    getBoundary: (tokenIterator, start) ->
-      moveToBoundary = if start then (=> tokenIterator.stepBackward()) else (=> tokenIterator.stepForward())
-      moveFromBoundary = if start then (=> tokenIterator.stepForward()) else (=> tokenIterator.stepBackward())
-      boundarySequences = (
-        if start
-        then EquationRangeHandler.BEGIN_EQUATION_TOKEN_SEQUENCES
-        else EquationRangeHandler.END_EQUATION_TOKEN_SEQUENCES
-      )
-
-      currentToken = tokenIterator.getCurrentToken()
-      prevRow = tokenIterator.getCurrentTokenPosition().row
-      # TODO: magic string? importing is hard though
-      boundaryCorrect = true
-      reasonCode = null
-      while LatexParsingContext.isType(currentToken, "equation")
-        currentToken = moveToBoundary()
-        if not currentToken?
-          boundaryCorrect = false
-          reasonCode = "js.math_preview.error.document_end"
-          break
-        currentRow = tokenIterator.getCurrentTokenPosition().row
-        # Empty string always means that equation state is popped from state stack.
-        # Unfortunately, empty string is not tokenized at all, and TokenIterator
-        # just skips it altogether, so we have to handle this manually here.
-        if Math.abs(currentRow - prevRow) > 1
-          boundaryCorrect = false
-          reasonCode = "js.math_preview.error.empty_line"
-          break
-        prevRow = currentRow
-
-      if currentToken? and LatexParsingContext.isType(currentToken, "error")
-        boundaryCorrect = false
-        reasonCode = "js.math_preview.error.whitespace_line"
-
-      moveFromBoundary()
-
-      { row: curTokenRow, column: curTokenColumn } = tokenIterator.getCurrentTokenPosition()
-      curTokenLength = tokenIterator.getCurrentToken().value.length
-      return {
-        correct: boundaryCorrect
-        reason: reasonCode
-        row: curTokenRow
-        column: curTokenColumn + (if start then 0 else curTokenLength)
-      }
+    getBoundary: (session, row, column, start) ->
+      summand = if start then -1 else 1
+      curIndex = session.doc.positionToIndex({ row, column })
+      { row: curRow, column: curColumn } = session.doc.indexToPosition(curIndex)
```

Comment:
What's the point of this trick? position to index and immediately index to position?